query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java

```diff
@@ -64,19 +71,27 @@ public Void apply(EmbeddedCacheManager embeddedCacheManager) {
       if (keys == null || keys.isEmpty()) {
          preIndex(cache, indexUpdater, notifier);
          MassIndexerProgressState progressState = new MassIndexerProgressState(notifier);
+
          if (!skipIndex) {
             try (Stream<CacheEntry<Object, Object>> stream = reindexCache.getAdvancedCache().withFlags(Flag.CACHE_MODE_LOCAL)
                   .cacheEntrySet().stream()) {
-               stream.forEach(entry -> {
-                  Object key = entry.getKey();
-                  Object value = valueDataConversion.extractIndexable(entry.getValue(), javaEmbeddedEntities);
-                  int segment = keyPartitioner.getSegment(key);
-
-                  if (value != null && indexUpdater.typeIsIndexed(value, indexedTypes)) {
-                     progressState.addItem(key, value,
-                           indexUpdater.updateIndex(key, value, segment));
-                  }
-               });
+               Set<UpdateRecord> records = stream.map(entry -> {
+                        Object key = entry.getKey();
+                        Object value = valueDataConversion.extractIndexable(entry.getValue(), javaEmbeddedEntities);
+                        int segment = keyPartitioner.getSegment(key);
+
+                        if (value != null && indexUpdater.typeIsIndexed(value, indexedTypes)) {
+                           return new UpdateRecord(key, value, segment);
+                        }
+                        return new UpdateRecord(null, null, -1);
+                     }).filter(record -> record.key != null)
+                     .collect(Collectors.toSet());
+
+               join(CompletionStages.performConcurrently(records, 100, Schedulers.from(new WithinThreadExecutor()), record -> {
```

Comment:
```suggestion
               Iterator<UpdateRecord> records = stream.map(entry -> {
                        Object key = entry.getKey();
                        Object value = valueDataConversion.extractIndexable(entry.getValue(), javaEmbeddedEntities);
                        int segment = keyPartitioner.getSegment(key);

                        if (value != null && indexUpdater.typeIsIndexed(value, indexedTypes)) {
                           return new UpdateRecord(key, value, segment);
                        }
                        return new UpdateRecord(null, null, -1);
                     }).filter(record -> record.key != null)
                     .iterator();

               join(CompletionStages.performConcurrently(() -> records, 100, Schedulers.from(new WithinThreadExecutor()), record -> {
```

This way we don't have to keep another copy of all entries in the cache and instead pull them as necessary.

Reply:
Right! Nice suggestion. Thanks